### PR TITLE
Reuse transfers in black-box scenarios

### PR DESCRIPTION
**Pull request by replaceafill**
Connects to https://github.com/archivematica/Issues/issues/622

### CODE REVIEW

File: features/steps/black_box_steps.py [hunk 1]
```diff
@@ -272,18 +272,19 @@ def step_impl(context):
     )
     tree = etree.parse(context.aip_mets_location)
     premis_events = tree.findall(
-        'mets:amdSec/mets:techMD/mets:mdWrap[@MDTYPE="PREMIS:EVENT"]/'
-        "mets:xmlData/premis:event",
+        'mets:amdSec/mets:digiprovMD/mets:mdWrap[@MDTYPE="PREMIS:EVENT"]/'
```
Comment (ross-spencer) [inline features/steps/black_box_steps.py]: Hi @replaceafill can you explain a little more about this change - is it driven by a chance in Archivematica?

Comment (replaceafill) [reply features/steps/black_box_steps.py]: @ross-spencer After talking to Evelyn I realized it was a false positive in the test. The `findall` was always returning an empty list because the `<mets:mdWrap MDTYPE="PREMIS:EVENT">` element has always been contained in the `<mets:digiprovMD>` element and not in the `<mets:techMD>` element.

That's also the reason why I added the assertion:

```
    assert premis_events, "Expected PREMIS events in the METS file"
```

a couple of lines below.

Comment (ross-spencer) [reply features/steps/black_box_steps.py]: Ahh! Cool. That makes sense. Thanks Douglas!

File: features/steps/utils.py [hunk 2]
```diff
@@ -283,186 +283,220 @@ def unzip(zip_path):
...
+    am.am_api_key = api_client_config["api_key"]
     return am
 
 
-def browse_default_ts_location(context, browse_path=None):
+def is_invalid_api_response(response):
+    return response is None or isinstance(response, int)
+
+
+def call_api_endpoint(
```
Comment (ross-spencer) [inline features/steps/utils.py]: I like this function! :+1:

File: features/steps/utils.py [hunk 3]
```diff
@@ -283,186 +283,220 @@ def unzip(zip_path):
...
+        error_message="Cannot list storage locations",
+    )
+    for location_object in response["objects"]:
+        if (
+            location_object.get("description") == ""
+            and location_object.get("enabled") is True
+            and location_object.get("relative_path").endswith("home")
+            and location_object.get("purpose") == "TS"
+        ):
+            return location_object.get("uuid")
```
Comment (ross-spencer) [inline features/steps/utils.py]: `EnvironmentError` is still being raised in `get_default_ss_pipeline(...)` is there a specific rationale for doing that there when we interact with the API, but not in this function, and the one above?

File: features/steps/utils.py [hunk 4]
```diff
@@ -283,186 +283,220 @@ def unzip(zip_path):
...
+def check_unapproved_transfers_with_same_directory_name(
+    api_clients_config, directory_name
+):
+    am = configure_am_client(api_clients_config["am"])
+    response = call_api_endpoint(
+        endpoint=am.unapproved_transfers,
+        warning_message="Cannot get unapproved tansfers",
+        error_message="Cannot get unapproved tansfers",
+    )
+    for transfer in response["results"]:
```
Comment (ross-spencer) [inline features/steps/utils.py]: How likely is it we will see a `TypeError` here? Now `call_api_endpoint(...)` exists, I wonder if there is any benefit guaranteeing the return of an empty dict from it to make failures more graceful? I think a lot of the tests would also handle that quite well if we are returning empty dicts and testing for `None`.

File: features/steps/utils.py [hunk 5]
```diff
@@ -283,186 +283,220 @@ def unzip(zip_path):
...
         raise environment.EnvironmentError(
             "Cannot start reingest of AIP {} because there is a "
             "unapproved transfer with the same directory name in "
-            "watchedDirectories/activeTransfers/standardTransfer".format(
-                context.sip_uuid
-            )
+            "watchedDirectories/activeTransfers/standardTransfer".format(sip_uuid)
         )
-    am = configure_ss_client(context)
+    am = configure_ss_client(api_clients_config["ss"])
```
Comment (ross-spencer) [inline features/steps/utils.py]: Rather than `"ss"`  and `"am"` below, these might read well if placed in constants in this script, or somewhere like the `amuser` as `storage_service` and `archivematica`.

File: features/steps/utils.py [hunk 6]
```diff
@@ -576,3 +610,110 @@ def get_filesec_files(tree, use=None, nsmap={}):
...
+    except environment.EnvironmentError as err:
+        assert False, "Error starting transfer: {}".format(err)
+
+
+def wait_for_transfer(api_clients_config, transfer_uuid):
+    status = None
+    resp = None
+    try:
+        while status not in ("COMPLETE", "FAILED") or (
+            status == "COMPLETE" and resp.get("sip_uuid") is None
```
Comment (ross-spencer) [inline features/steps/utils.py]: Note to self to add this information to the API wiki to help users trying to make use of this call.

File: features/steps/utils.py [hunk 7]
```diff
@@ -576,3 +610,110 @@ def get_filesec_files(tree, use=None, nsmap={}):
...
+    extracted_aip_dir = extract_aip(api_clients_config, sip_uuid)
+    aip_mets_location = get_aip_mets_location(extracted_aip_dir, sip_uuid)
+    return {
+        "sip_uuid": sip_uuid,
+        "extracted_aip_dir": extracted_aip_dir,
+        "aip_mets_location": aip_mets_location,
+    }
+
+
+def create_sample_transfer(api_clients_config, sample_transfer_path):
```
Comment (ross-spencer) [inline features/steps/utils.py]: :sunglasses:

File: features/black_box/create-aip.feature [hunk 8]
```diff
@@ -1,4 +1,4 @@
-@black-box
+@black-box @black-box-create-aip
```
[the remark this answers is not in the document]
Comment (replaceafill) [reply features/black_box/create-aip.feature]: I guess I added the tags to have more control on what's run, but I agree maybe other `behave` options like `--include` are more convenient. My decision was heavily influenced by the convenience of the docker-compose environment I use where `make test-at-behave TAGS=...` is available.

Maybe I should be running the tests differently, and a Makefile in this repository makes sense. We should try to sync those new make features with the am.git repository.

There's already some [documentation about the old tags](https://github.com/artefactual-labs/archivematica-acceptance-tests/blob/master/docs/developer-documentation.rst#how-are-the-tests-run-in-practice) we could adjust and expand for the index.

Maybe we should file a new issue (or look for one) and move this discussion there? In the meantime I can take the new tags out.

File: features/black_box/create-aip.feature [hunk 9]
```diff
@@ -40,7 +40,7 @@ Background: The storage service is configured with a transfer source that can se
     When the AIP is downloaded
     Then the physical structMap of the AIP METS accurately reflects the physical layout of the AIP
 
-  Scenario: Every object contains has a UUID
+  Scenario: Every object has a UUID
```
[the remark this answers is not in the document]
Comment (replaceafill) [reply features/black_box/create-aip.feature]: Ah! Thanks for your explanation. I'll include the fix for the `UUID` lookup in this PR.

File: features/black_box/create-aip.feature (hunk 8 above)
[the remark this answers is not in the document]
Comment (ross-spencer) [reply features/black_box/create-aip.feature]: A new issue sounds great @replaceafill, sorry, didn't mean to imply lumbering you with this new task :)

File: features/environment.py [hunk 10]
```diff
@@ -171,7 +170,18 @@ def before_scenario(context, scenario):
...
+            "url": context.am_user.ss_url.rstrip("/"),
+            "username": context.am_user.ss_username,
+            "api_key": context.am_user.browser.ss_api_key,
+        },
+        "am": {
+            "url": context.am_user.am_url.rstrip("/"),
+            "username": context.am_user.am_username,
+            "api_key": context.am_user.am_api_key,
+        },
+    }
```
[the remark this answers is not in the document]
Comment (replaceafill) [reply features/environment.py]: This is a mistake on my part for relying on the `am.git` make rule that conveniently [passes several user settings](https://github.com/artefactual-labs/am/blob/master/compose/Makefile#L164-L176) and makes the driver configuration to work. I should probably start running the tests more isolated.

Could you please clarify this part:

> make sure that each test time is properly isolated

File: features/steps/black_box_steps.py [hunk 11]
```diff
@@ -9,125 +9,46 @@
...
-
-
 @given("an AIP has been created and stored")
 def step_impl(context):
-    context.execute_steps(
-        'Given the transfer "DemoTransferCSV" is started with the'
-        " automatedProcessingMCP processing configuration.\n"
-        "When the Transfer is COMPLETE\n"
-        "And the Ingest is COMPLETE\n"
+    sample_transfer_path = "SampleTransfers/DemoTransferCSV"
```
[the remark this answers is not in the document]
Comment (replaceafill) [reply features/steps/black_box_steps.py]: :+1:

File: features/steps/black_box_steps.py [hunk 12]
```diff
@@ -158,83 +81,92 @@ def step_impl(context):
...
             cached_directories[file_dir] = file_dir_browse_result
         assert file_name in cached_directories[file_dir]["entries"]
 
 
 @then("the AIP contains a file called README.html")
 def step_impl(context):
-    readme_file = "{}-{}/data/README.html".format(
-        context.transfer_name, context.sip_uuid
+    readme_file = utils.get_aip_file_location(
+        context.current_transfer["extracted_aip_dir"], "data/README.html"
```
[the remark this answers is not in the document]
Comment (replaceafill) [reply features/steps/black_box_steps.py]: :+1:

File: features/steps/utils.py (hunk 5 above)
Comment (replaceafill) [reply features/steps/utils.py]: I think it would have to be from the `environment` module since that's what the `utils` module has access to. Sorry, I'm not sure I follow this recommendation. Do you mean something like this:

```
# environment.py
STORAGE_SERVICE_API_CONFIG_KEY = 'storage_service'
ARCHIVEMATICA_API_CONFIG_KEY = 'archivematica'
...
def before_scenario(context, scenario):
    ...
    context.api_clients_config = {
        STORAGE_SERVICE_API_CONFIG_KEY: ...,
        ARCHIVEMATICA_API_CONFIG_KEY: ...,
    }

# utils.py
from environment import 
...
    am = configure_ss_client(api_clients_config[STORAGE_SERVICE_API_CONFIG_KEY])
...
```

File: features/environment.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ross-spencer) [reply features/environment.py]: > make sure that each test time is properly isolated

I think that's a typo. I was just noting that it looked like it might only be a small amount of work to continue to keep the black box tests isolated from the Selenium ones. At least when we initialize them.

File: features/steps/utils.py (hunk 5 above)
Comment (ross-spencer) [reply features/steps/utils.py]: That looks good @replaceafill, would that be okay to do? I feel like it's just a little nicer than using quote encoded strings each time we need to retrieve the config.

File: features/steps/utils.py (hunk 6 above)
Comment (ross-spencer) [reply features/steps/utils.py]: I have added this to the wiki: https://wiki.archivematica.org/Archivematica_API#Status

File: features/environment.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ross-spencer) [reply features/environment.py]: I've updated the branch, and it's looking better. I'm still getting: 'HOOK-ERROR in after_scenario: AttributeError: 'NoneType' object has no attribute 'window_handles''

We can check `if not None` or `except` the `AttributeError` in the teardown to avoid this I think.

File: features/steps/black_box_steps.py [hunk 13]
```diff
@@ -179,11 +179,12 @@ def step_impl(context):
         amdsec = tree.find(
             'mets:amdSec[@ID="{}"]'.format(amdsec_id), namespaces=context.mets_nsmap
         )
-        object_uuid = amdsec.findtext(
+        object_uuid = amdsec.xpath(
             "mets:techMD/mets:mdWrap/mets:xmlData/premis3:object/"
-            "premis3:objectIdentifier/premis3:objectIdentifierValue",
+            'premis3:objectIdentifier/premis3:objectIdentifierType[text()="UUID"]/'
+            "../premis3:objectIdentifierValue",
```
Comment (ross-spencer) [inline features/steps/black_box_steps.py]: :+1:

File: features/environment.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (replaceafill) [reply features/environment.py]: Ah yes! I set it up in a separate virtual environment and saw the same error. I'll try to fix it.

File: features/steps/utils.py (hunk 5 above)
Comment (replaceafill) [reply features/steps/utils.py]: :+1:

File: features/environment.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (replaceafill) [reply features/environment.py]: Just for reference, after https://github.com/artefactual-labs/archivematica-acceptance-tests/pull/140/commits/36b14f201eb32b9d3599302d5bb3d80859e1421d the `black-box` tests can be run like this:

```
$ behave --no-skipped -v --stop -D am_url=http://127.0.0.1:62080/ -D am_username=test -D am_api_key=test -D ss_url=http://127.0.0.1:62081/ -D ss_username=test -D ss_api_key=test --tags=black-box
```

File: features/black_box/create-aip.feature (hunk 8 above)
[the remark this answers is not in the document]
Comment (replaceafill) [reply features/black_box/create-aip.feature]: I filed https://github.com/archivematica/Issues/issues/648 for this and removed the new tags.

**Comment by replaceafill** [comment]
@ross-spencer I think is ready for another review.

File: features/steps/black_box_steps.py [hunk 14]
```diff
@@ -9,125 +9,58 @@
...
-        "And the Ingest is COMPLETE\n"
+def format_original_files_error(transfer):
+    return 'The {} file does not contain any "original" files in its fileSec'.format(
+        transfer["aip_mets_location"]
     )
 
 
-@given("an AIP has been reingested")
-def step_impl(context):
-    context.execute_steps(
```
Comment (kenoir) [inline features/steps/black_box_steps.py]: 👏 Great to get rid of these `context.execute_steps`! Calling steps from step files seems like a bit of an anti-pattern.

File: features/environment.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ross-spencer) [reply features/environment.py]: That's great - those tests with that command run exactly as intended! 
```
3 features passed, 0 failed, 20 skipped
18 scenarios passed, 0 failed, 52 skipped
54 steps passed, 0 failed, 558 skipped, 0 undefined
Took 4m33.315s
```

Nice work!
